### PR TITLE
UNR-2129 Removed the Sender from the Enforcer

### PR DESCRIPTION
**Pull request by aleximprobable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Removed Sender from Enforcer. Now Enforcer returns an array of requests to be processed by the sender.

#### Release note
No release note update

#### Tests
Zoning tests gym should work as before the change.

#### Documentation
No doc update.

#### Primary reviewers
@m-samiec @samiwh

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp [hunk 1]
```diff
@@ -117,7 +95,7 @@ void SpatialLoadBalanceEnforcer::ProcessQueuedAclAssignmentRequests()
 			// TODO(zoning): Not sure whether this should be possible or not. Remove if we don't see the warning again.
 			UE_LOG(LogSpatialLoadBalanceEnforcer, Warning, TEXT("(%s) Entity without AuthIntent component will not be processed. EntityId: %lld"), *WorkerId, Request.EntityId);
 			CompletedRequests.Add(Request.EntityId);
-			return;
+			return PendingRequests;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: Driveby - this should be continue - see https://github.com/spatialos/UnrealGDK/pull/1555

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp [hunk 2]
```diff
@@ -104,8 +80,10 @@ void SpatialLoadBalanceEnforcer::QueueAclAssignmentRequest(const Worker_EntityId
 	}
 }
 
-void SpatialLoadBalanceEnforcer::ProcessQueuedAclAssignmentRequests()
+TArray<SpatialLoadBalanceEnforcer::AclWriteAuthorityRequest> SpatialLoadBalanceEnforcer::ProcessQueuedAclAssignmentRequests()
 {
+	TArray<SpatialLoadBalanceEnforcer::AclWriteAuthorityRequest> PendingRequests;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: To avoid the copy, please pass this Array in as a reference, or MoveTemp it out.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp (hunk 1 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: Yes, this should be `continue`, so changing it to `return PendingRequests` should fix the issue

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp (hunk 2 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: As I remember, Return-Value optimisation should take care of that automatically.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp (hunk 1 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: But if you return you aren't continuing...

Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: All right :D I'll `continue`

Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: Addressed

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h [hunk 3]
```diff
@@ -3,32 +3,36 @@
...
-class USpatialSender;
 class USpatialStaticComponentView;
 
 class SpatialLoadBalanceEnforcer
 {
 public:
-	SpatialLoadBalanceEnforcer();
+	struct AclWriteAuthorityRequest
+	{
+		Worker_EntityId EntityId = 0;
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h]: For a struct I don't think this is really necessary.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h]: You can zero init an instance if you need to. Otherwise you plan on setting every value anyway.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 4]
```diff
@@ -1555,7 +1554,10 @@ void USpatialNetDriver::TickDispatch(float DeltaTime)
 
 		if (LoadBalanceEnforcer.IsValid())
 		{
-			LoadBalanceEnforcer->Tick();
+			for(const auto& elem : LoadBalanceEnforcer->ProcessQueuedAclAssignmentRequests())
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Not sure if ProcessQueuedAclAssignmentRequests is now the correct function name for this? Maybe it is, it just made me think.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h [hunk 5]
```diff
@@ -3,32 +3,36 @@
...
+		FString OwningWorkerId;
+	};
 
-	void Init(const FString &InWorkerId, USpatialStaticComponentView* InStaticComponentView, USpatialSender* InSpatialSender, SpatialVirtualWorkerTranslator* InVirtualWorkerTranslator);
-	void Tick();
+	SpatialLoadBalanceEnforcer(const FString &InWorkerId, USpatialStaticComponentView* InStaticComponentView, SpatialVirtualWorkerTranslator* InVirtualWorkerTranslator);
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h]: You might have to also change GetPhysicalWorkerForVirtualWorker to be const as well

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 4 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: I agree, there might be a better name. It still does some processing internally though, so I kept it as is for now. What name could you suggest?

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h (hunk 5 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h]: Made `GetPhysicalWorkerForVirtualWorker` const

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h [hunk 6]
```diff
@@ -3,33 +3,37 @@
...
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h]: Pointer arguments in the constructor can be const?

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h (hunk 3 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h]: It isn't really necessary, but IMO being explicit is nice.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 7]
```diff
@@ -1553,7 +1552,10 @@ void USpatialNetDriver::TickDispatch(float DeltaTime)
...
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]:
```suggestion
			for(const auto& Elem : LoadBalanceEnforcer->ProcessQueuedAclAssignmentRequests())
```

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialStaticComponentView.h [hunk 8]
```diff
@@ -20,31 +20,38 @@ class SPATIALGDK_API USpatialStaticComponentView : public UObject
...
+	bool HasComponent(Worker_EntityId EntityId, Worker_ComponentId ComponentId) const;
 
 	void OnAddComponent(const Worker_AddComponentOp& Op);
 	void OnRemoveComponent(const Worker_RemoveComponentOp& Op);
 	void OnRemoveEntity(Worker_EntityId EntityId);
 	void OnComponentUpdate(const Worker_ComponentUpdateOp& Op);
 	void OnAuthorityChange(const Worker_AuthorityChangeOp& Op);
 
-	void GetEntityIds(TArray<Worker_EntityId_Key>& EntityIds) { EntityComponentMap.GetKeys(EntityIds); }
+	void GetEntityIds(TArray<Worker_EntityId_Key>& EntityIds) const { EntityComponentMap.GetKeys(EntityIds); }
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialStaticComponentView.h]:
```suggestion
	void GetEntityIds(TArray<Worker_EntityId_Key>& OutEntityIds) const { EntityComponentMap.GetKeys(OutEntityIds); }
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 7 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Applied.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h (hunk 6 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h]: Changed.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h [hunk 9]
```diff
@@ -4,35 +4,39 @@
...
 public:
-	SpatialLoadBalanceEnforcer();
+	struct AclWriteAuthorityRequest
+	{
+		Worker_EntityId EntityId = 0;
+		FString OwningWorkerId;
+	};
 
-	void Init(const PhysicalWorkerName &InWorkerId, USpatialStaticComponentView* InStaticComponentView, USpatialSender* InSpatialSender, SpatialVirtualWorkerTranslator* InVirtualWorkerTranslator);
+	SpatialLoadBalanceEnforcer(const FString &InWorkerId, const USpatialStaticComponentView* InStaticComponentView, const SpatialVirtualWorkerTranslator* InVirtualWorkerTranslator);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialLoadBalanceEnforcer.h]:
```suggestion
	SpatialLoadBalanceEnforcer(const PhysicalWorkerName &InWorkerId, const USpatialStaticComponentView* InStaticComponentView, const SpatialVirtualWorkerTranslator* InVirtualWorkerTranslator);	
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp [hunk 10]
```diff
@@ -12,38 +11,17 @@ DEFINE_LOG_CATEGORY(LogSpatialLoadBalanceEnforcer);
...
-	, Sender(nullptr)
-	, VirtualWorkerTranslator(nullptr)
-{
-}
-
-void SpatialLoadBalanceEnforcer::Init(const PhysicalWorkerName &InWorkerId,
-	USpatialStaticComponentView* InStaticComponentView,
-	USpatialSender* InSpatialSender,
-	SpatialVirtualWorkerTranslator* InVirtualWorkerTranslator)
+SpatialLoadBalanceEnforcer::SpatialLoadBalanceEnforcer(const FString &InWorkerId, const USpatialStaticComponentView* InStaticComponentView, const SpatialVirtualWorkerTranslator* InVirtualWorkerTranslator)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]:
```suggestion
SpatialLoadBalanceEnforcer::SpatialLoadBalanceEnforcer(const FString& InWorkerId, const USpatialStaticComponentView* InStaticComponentView, const SpatialVirtualWorkerTranslator* InVirtualWorkerTranslator)
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp [hunk 11]
```diff
@@ -53,7 +53,7 @@ void SpatialVirtualWorkerTranslator::AddVirtualWorkerIds(const TSet<VirtualWorke
 	}
 }
 
-const PhysicalWorkerName* SpatialVirtualWorkerTranslator::GetPhysicalWorkerForVirtualWorker(VirtualWorkerId id)
+const PhysicalWorkerName* SpatialVirtualWorkerTranslator::GetPhysicalWorkerForVirtualWorker(VirtualWorkerId id) const
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp]:
```suggestion
const PhysicalWorkerName* SpatialVirtualWorkerTranslator::GetPhysicalWorkerForVirtualWorker(VirtualWorkerId Id) const
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp [hunk 12]
```diff
@@ -53,7 +53,7 @@ void SpatialVirtualWorkerTranslator::AddVirtualWorkerIds(const TSet<VirtualWorke
...
 {
 	return VirtualToPhysicalWorkerMapping.Find(id);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp]:
```suggestion
	return VirtualToPhysicalWorkerMapping.Find(Id);
```

Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp]: Updated

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp (hunk 11 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp]: Updated

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp (hunk 10 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp]: Updated
